docs/enhancements/clusterpool-inventory.md

```diff
@@ -0,0 +1,186 @@
+
+But that doesn't work for some cloud providers such a vSphere, OpenStack and Baremetals.
+
+In such cases, one solution is to configure DNS manually and then create your ClusterDeployment with the right `Name` and `BaseDomain` so that the assembled hostnames match those entries. This is fine if you're creating ClusterDeployments by hand, but breaks down if ClusterDeployment names are being generated with random slugs, as is the case with ClusterPools.
+
+## Proposal
+### Summary
+Allow ClusterPool to accept an inventory of ClusterDeploymentCustomizations that will have install config json patches to be used when generating ClusterDeployments for the pool.
+
+### `ClusterPool.Spec.Inventory`
+Add a field to `ClusterPool.Spec` called `Inventory`.
+
+It has an array of object references to ClusterDeploymentCustomization custom resource. This CR will have a json patch (RFC 6902) for default install config generated by clusterpool controller. The json patch will have cluster specific changes, for vSphere this corresponds to configured DNS hostnames
+
+```yaml
+spec:
+  inventory:
+    ClusterDeploymentCustomizations:
+      - name: foo-cluster-deployment-customization
+      - name: bar-cluster-deployment-customization
+      - name: baz-cluster-deployment-customization
+```
+
+and ClusterDeploymentCustomization CR will look like
+```yaml
+apiVersion: v1
+kind: ClusterDeploymentCustomization
+metadata:
+  name: foo-cluster-deployment-customization
+  namespace: my-project
+spec:
+  installConfigPatches:
+    - op: replace
+      path: metadata/name
+      value: foo
+status:
+  clusterDeploymentRef: 
+    name: foo
+    namespace: foo-namespace
+  conditions:
+    - lastProbeTime: "2020-11-05T14:49:26Z"
+      lastTransitionTime: "2020-11-05T14:49:26Z"
+      message: Currently in use by cluster deployment foo of clusterpool foo-pool
+      reason: ClusterDeploymentCustomizationInUse
+      status: "false"
+      type: Available
+```
+
+if DNS is configured with the name `foo`, ClusterDeploymentCustomization.spec.installConfigPatches content to patch vSphere install config will be as follows
+```yaml
+spec:
+  installConfigPatches:
+    - op: replace
+      path: metadata/name
+      value: foo
+```
+
+When adding a ClusterDeployment, if such an `Inventory` is present, ClusterPool controller will:
+- Load up the inventory list.
+- For each reference it will do a GET to fetch ClusterDeploymentCustomization and check if the `Available` condition is true and `.status.clusterDeploymentRef` is nil.
+- For the first ClusterDeploymentCustomization that is available, it will use the patches in the `spec.installConfigPatchs` field to update the default install config generated by clusterpool. The patches will be applied in the order listed in the inventory. It will also update the status with reference to the cluster deployment that is using it and update the `Available` condition to false.
```

Comment:
```suggestion
- For the first ClusterDeploymentCustomization that is available, it will use the patches in the `spec.installConfigPatches` field to update the default install config generated by clusterpool. The patches will be applied in the order listed in the inventory. It will also update the status with reference to the cluster deployment that is using it and update the `Available` condition to false.
```